src/lib/parsers/ChatInputCommandArgs.ts

```diff
@@ -0,0 +1,626 @@
+		const result = await this.peekResult(options);
+		return result.unwrapRaw();
+	}
+
+	/**
+	 * Saves the current state into the stack following a FILO strategy (first-in, last-out).
+	 * @see Args#restore
+	 */
+	public save(): void {
+		this.states.push(this.parser.save());
+	}
+
+	/**
+	 * Restores the previously saved state from the stack.
+	 * @see Args#save
+	 */
+	public restore(): void {
+		if (this.states.length !== 0) this.parser.restore(this.states.pop()!);
+	}
+
+	/**
+	 * Whether all arguments have been consumed.
+	 */
+	public get finished(): boolean {
+		return this.parser.finished;
+	}
+
+	/**
+	 * Defines the `JSON.stringify` override.
+	 */
+	public toJSON(): ArgsJson {
+		return { message: this.interaction, command: this.command, commandContext: this.commandContext };
+	}
+
+	protected unavailableArgument<T>(type: string | IArgument<T>): Result.Err<UserError> {
+		const name = typeof type === 'string' ? type : type.name;
+		return Result.err(
+			new UserError({
+				identifier: Identifiers.ArgsUnavailable,
+				message: `The argument "${name}" was not found.`,
+				context: { name, ...this.toJSON() }
+			})
+		);
+	}
+
+	protected missingArguments(): Result.Err<UserError> {
+		return Result.err(new UserError({ identifier: Identifiers.ArgsMissing, message: 'There are no more arguments.', context: this.toJSON() }));
+	}
+
+	/**
+	 * Resolves an argument.
+	 * @param arg The argument name or {@link IArgument} instance.
+	 */
+	private resolveArgument<T>(arg: keyof ArgType | IArgument<T>): IArgument<T> | undefined {
+		if (typeof arg === 'object') return arg;
+		return container.stores.get('arguments').get(arg as string) as IArgument<T> | undefined;
+	}
+}
+
+export interface ArgsJson {
+	message: Message | AnyInteraction;
+	command: Command;
+	commandContext: Record<PropertyKey, unknown>;
+}
+
+export interface ArgType {
+	boolean: boolean;
+	channel: ChannelTypes;
+	date: Date;
+	dmChannel: DMChannel;
+	emoji: EmojiObject;
+	float: number;
+	guildCategoryChannel: CategoryChannel;
+	guildChannel: GuildBasedChannelTypes;
+	guildNewsChannel: NewsChannel;
+	guildNewsThreadChannel: ThreadChannel & { type: ChannelType.AnnouncementThread; parent: NewsChannel | null };
+	guildPrivateThreadChannel: ThreadChannel & { type: ChannelType.PrivateThread; parent: TextChannel | null };
+	guildPublicThreadChannel: ThreadChannel & { type: ChannelType.PublicThread; parent: TextChannel | null };
+	guildStageVoiceChannel: StageChannel;
+	guildTextChannel: TextChannel;
+	guildThreadChannel: ThreadChannel;
+	guildVoiceChannel: VoiceChannel;
+	hyperlink: URL;
+	integer: number;
+	member: GuildMember;
+	message: Message;
+	number: number;
+	role: Role;
+	string: string;
+	url: URL;
+	user: User;
+	enum: string;
+}
+
+/**
+ * The callback used for {@link Args.nextMaybe} and {@link Args.next}.
+ */
+export interface ArgsNextCallback<T> {
+	/**
+	 * The value to be mapped.
+	 */
+	(value: CommandInteraction): Option<T>;
+}
```

Comment:
Ideally this would be on the base class with a generic. That said, why is the value an interaction?

Reply:
Actually, this is unused in ChatInputCommandArgs since it doesn't have next/nextMaybe. I'll leave it just in MessageArgs, but let me know if I should still move it to the base class.

Reply:
if it stays only for message args, then move it there only! But anything that both args should implement should always be present and exported from the base class